src/sources/exec.rs

```diff
@@ -0,0 +1,843 @@
+        if !arguments.is_empty() {
+            command.args(arguments);
+        }
+    }
+
+    command
+}
+
+fn create_event(
+    config: &ExecConfig,
+    line: Bytes,
+    data_stream: &Option<String>,
+    pid: Option<u32>,
+    exit_status: Option<i32>,
+    exec_duration_millis: &Option<u128>,
+) -> Event {
+    emit!(ExecEventReceived {
+        command: config.command.as_str(),
+        byte_size: line.len(),
+    });
+
+    let mut event = Event::from(line);
+    let log_event = event.as_mut_log();
+
+    // Add source type
+    log_event.insert(log_schema().source_type_key(), Bytes::from(EXEC));
+
+    // Add data stream of stdin or stderr (if needed)
+    if let Some(data_stream) = data_stream {
+        log_event.insert(DATA_STREAM_KEY, data_stream.clone());
+    }
+
+    // Add pid (if needed)
+    if let Some(pid) = pid {
+        log_event.insert(PID_KEY, pid as i64);
+    }
+
+    // Add exit status (if needed)
+    if let Some(exit_status) = exit_status {
+        log_event.insert(EXIT_STATUS_KEY, exit_status as i64);
+    }
+
+    // Add exec duration millis (if needed)
+    if let Some(exec_duration_millis) = exec_duration_millis {
+        log_event.insert(EXEC_DURATION_MILLIS_KEY, *exec_duration_millis as i64);
+    }
+
+    // Add hostname (if needed)
+    if let Some(hostname) = config.hostname.clone() {
+        log_event.insert(log_schema().host_key(), hostname);
+    }
+
+    // Add command
+    log_event.insert(COMMAND_KEY, config.command.clone());
+
+    // Add arguments
+    log_event.insert(ARGUMENTS_KEY, config.arguments.clone());
+
+    event
+}
+
+fn spawn_reader_thread<R: 'static + AsyncRead + Unpin + std::marker::Send>(
+    reader: BufReader<R>,
+    shutdown: ShutdownSignal,
+    event_per_line: bool,
+    buf_size: usize,
+    stream: &'static str,
+    mut sender: Sender<(String, &'static str)>,
+) {
+    // Start the green background thread for collecting
+    Box::pin(tokio::spawn(async move {
+        info!("Start capturing {} command output.", stream);
+
+        let mut buffer: Vec<u8> = Vec::new();
+
+        let mut reader = reader.allow_read_until(shutdown.clone().map(|_| ()));
+
+        // Read one byte at a time so we don't block waiting for the buffer to
+        // fill up e.g. If a line filled up half the buffer we would not know about
+        // it until the buffer is returned. We could increase the read buffer if we are
+        // willing to accept some blocking waiting for it to fill up. I'm not sure if
+        // the underlying BufReader will read in larger chunks or not.
+        // Possibly could look into https://crates.io/crates/fixed-buffer-tokio but
+        // I don't know much about this library.
+        let mut read_buffer = [0_u8; 1];
+
+        // Not using the shutdown signal in this method so we need to shutdown
+        // in other methods to end the reading
+        while let Ok(bytes_read) = reader.read(&mut read_buffer).await {
```

Comment:
It should be easy enough to change this to use lines() but before I change it can I clarify if we should also drop the maximum_buffer_size? I think if we use lines() or read_until \n then we cannot control the size of the string or bytes returned. I suppose we could chop the lines/bytes returned but it seems too late at that point.

Reply:
Right, if we switch to `lines()` we should drop `maximum_buffer_size` too. Given that the `stdin` source already has a precedent for just doing `lines()`, I'm ok with that for the initial implementation here. We could follow up later to re-add support for limiting the buffer size, which I think could be done with only slightly more complexity via `reader.by_ref().take(buf_size).read_until('\n', &mut read_buffer`.

Reply:
Oh that is really nice! I don't mind giving that implementation a shot and then switching to lines if I can't get it to work (or if you guys are unhappy with it). I really like how event_per_line and maximum_buffer_size work really well together for both streaming and scheduled so if we can get it to work using the method you suggested that would be great. If you want lines in order to keep the event_per_line applicable to both scheduled and streaming we could implement a poor mans maximum_buffer_size called maximum_lines. Does the above sound ok to you?

Reply:
I really liked your suggestion of using `take(buf_size)` and so I have implemented it. I had to `async_buf_read.rs` module as the `async_read.rs` one only implemented `AsyncRead` and not `AsyncBufRead` but other than that I think it turned out quite nicely.

Since you were OK with cutting off a UTF-8 character I removed the UTF validation logic and simply send the raw bytes. @bruceg suggested a really nice improvement to the `buffer_to_string` method using `Utf8Error::valid_up_to` so if we ever want to add some encoding logic back in we could conditionally add it back using the logic he suggested.